benchmarks/src/jmh/java/io/grpc/netty/InboundHeadersBenchmark.java

```diff
@@ -0,0 +1,202 @@
+    requestHeaders[i++] = of(":path");
+    requestHeaders[i++] = of("/google.pubsub.v2.PublisherService/CreateTopic");
+    requestHeaders[i++] = of(":authority");
+    requestHeaders[i++] = of("pubsub.googleapis.com");
+    requestHeaders[i++] = of("te");
+    requestHeaders[i++] = of("trailers");
+    requestHeaders[i++] = of("grpc-timeout");
+    requestHeaders[i++] = of("1S");
+    requestHeaders[i++] = of("content-type");
+    requestHeaders[i++] = of("application/grpc+proto");
+    requestHeaders[i++] = of("grpc-encoding");
+    requestHeaders[i++] = of("gzip");
+    requestHeaders[i++] = of("authorization");
+    requestHeaders[i] = of("Bearer y235.wef315yfh138vh31hv93hv8h3v");
+  }
+
+  private static void setupResponseHeaders() {
+    responseHeaders = new AsciiString[4];
+    int i = 0;
+    responseHeaders[i++] = of(":status");
+    responseHeaders[i++] = of("200");
+    responseHeaders[i++] = of("grpc-encoding");
+    responseHeaders[i] = of("gzip");
+  }
+
+  /**
+   * Checkstyle.
+   */
+  @Benchmark
+  @BenchmarkMode(Mode.AverageTime)
+  @OutputTimeUnit(TimeUnit.NANOSECONDS)
+  public void grpcHeaders_serverHandler(Blackhole bh) {
+    serverHandler(bh, new GrpcHttp2RequestHeaders(4));
+  }
+
+  /**
+   * Checkstyle.
+   */
+  @Benchmark
+  @BenchmarkMode(Mode.AverageTime)
+  @OutputTimeUnit(TimeUnit.NANOSECONDS)
+  public void defaultHeaders_serverHandler(Blackhole bh) {
+    serverHandler(bh, new DefaultHttp2Headers(true, 9));
+  }
+
+  /**
+   *  Checkstyle.
+   */
+  @Benchmark
+  @BenchmarkMode(Mode.AverageTime)
+  @OutputTimeUnit(TimeUnit.NANOSECONDS)
+  public void grpcHeaders_clientHandler(Blackhole bh) {
+    clientHandler(bh, new GrpcHttp2ResponseHeaders(2));
+  }
+
+  /**
+   * Checkstyle.
+   */
+  @Benchmark
+  @BenchmarkMode(Mode.AverageTime)
+  @OutputTimeUnit(TimeUnit.NANOSECONDS)
+  public void defaultHeaders_clientHandler(Blackhole bh) {
+    clientHandler(bh, new DefaultHttp2Headers(true, 2));
+  }
+
+  @CompilerControl(CompilerControl.Mode.INLINE)
+  private static void serverHandler(Blackhole bh, Http2Headers headers) {
+    for (int i = 0; i < requestHeaders.length; i += 2) {
+      bh.consume(headers.add(requestHeaders[i], requestHeaders[i + 1]));
+    }
+
+    // Sequence of headers accessed in NettyServerHandler
+    bh.consume(headers.get(TE_TRAILERS));
+    bh.consume(headers.get(CONTENT_TYPE_HEADER));
+    bh.consume(headers.method());
+    bh.consume(headers.get(CONTENT_TYPE_HEADER));
+    bh.consume(headers.path());
+
+    bh.consume(Utils.convertHeaders(headers));
+  }
+
+  @CompilerControl(CompilerControl.Mode.INLINE)
+  private static void clientHandler(Blackhole bh, Http2Headers headers) {
+    // NettyClientHandler does not directly access headers, but convert to Metadata immediately.
+
+    bh.consume(headers.add(responseHeaders[0], responseHeaders[1]));
+    bh.consume(headers.add(responseHeaders[2], responseHeaders[3]));
+
+    bh.consume(Utils.convertHeaders(headers));
+  }
+
+//  /**
```

Comment:
I wanted to keep it so in case of changes to the headers we can undocument it and re-evaluate the memory usage. I only commented it out, so that we don't have the JOL dependency.